python/phase-2.md

```diff
@@ -828,4 +829,168 @@ Notes:
\ No newline at end of file
+* We will, however, go ahead and abstract the "runner" so that we can make it pluggable/opt-out if the caller chooses
+
+## Additional API Concerns
+
+### Activities as Methods
+
+It is very normal to want to have activities as methods instead of top-level functions. For example, say you have:
+
+```python
+class ActivityClass:
+    def __init__(self, some_init_arg: int) -> None:
+        pass
+    
+    @activity.defn
+    def some_activity(self, arg1: str) -> List[str]:
+        return ["Yay"]
+```
+
+You cannot just `workflow.execute_activity(ActivityClass.some_activity, "foo")` because technically the first parameter
+is `self` there. But sometimes they may want to call
+`workflow.execute_activity(ActivityClass.some_activity, ActivityClass("init-arg1"), "foo")` and who are we to say they
+can't pass in the self arg explicitly?
+
+What do we do? Here are the notes/options (activity options left off of calls for brevity):
+
+* Cannot use something like `functools.partial(ActivityClass.some_activity, None)` because `partial` removes our
+  attributes which are set in the decorator
+* Don't want to init an instance of the class like this:
+
+    ```python
+    workflow.execute_activity(ActivityClass("init-arg").some_activity, "foo")
+    ```
+* Can create the activity class without initializing it like this:
+
+    ```python
+    activities = ActivityClass.__new__(ActivityClass)
+    workflow.execute_activity(activities.some_activity, "foo")
+    ```
+  This is nice because it doesn't require the Python SDK to care (and people have been doing this in class methods for
+  factories for a while now). We can even provide a helper that does basically this like so:
+
+    ```python
+    activities = workflow.proxy_activity_class(ActivityClass)
+    workflow.execute_activity(activities.some_activity, "foo")
+    ```
+* We can make it possible to pass `None` as the `self` parameter like this:
+
+    ```python
+    workflow.execute_activity(ActivityClass.some_activity, None, "foo")
+    ```
+  And then we could check for that being self and ignore it. However, due to Mypy not yet supporting all of `ParamSpec`,
+  that requires us to have a two-parameter overload version of `execute_activity` _and_ to make the first param optional
+  like so:
+  
+    ```python
+    @overload
+    async def execute_activity(
+        activity: Callable[[LocalParamType, LocalParamType2], LocalReturnType],
+        arg: Optional[LocalParamType],
+        arg2: LocalParamType2,
+        /,
+        *,
+    ```
+  This means if you had a real two-param activity, you'd now be able to have an optional first parameter. Also, I am not
+  yet confident how this will look when we remove all of these overloads once Mypy gets their `ParamSpec` support.
+* We could have a completely separate call like so:
+
+    ```python
+    workflow.execute_activity_method(ActivityClass.some_activity, "foo")
```

Comment:
I don't have a strong preference here, some other options you've mentioned are completely valid too.

This is the cleanest IMHO with the current limitations of python typings.

The proxy approach introduces new concepts that I think we should avoid, it also make goto definition harder (lessons learned from TS).

The other valid option is have the `method=True` flag but I feel like the signature would be more confusing with additional overloads.